pkg/ingester/index/index.go

```diff
@@ -1,42 +1,105 @@
-package ingester
+package index
 
 import (
 	"sort"
 	"sync"
+	"unsafe"
 
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/pkg/labels"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
 )
 
 const indexShards = 32
 
-type invertedIndex struct {
+// InvertedIndex implements a in-memory inverter index from label pairs to fingerprints.
+// It is sharded to reduce lock contention on writes.
+type InvertedIndex struct {
 	shards []indexShard
 }
 
-type indexShard struct {
-	mtx sync.RWMutex
-	idx map[model.LabelName]map[model.LabelValue][]model.Fingerprint // slice entries are sorted in fp order
-}
-
-func newInvertedIndex() *invertedIndex {
+// New returns an new InvertedIndex.
+func New() *InvertedIndex {
 	shards := make([]indexShard, indexShards)
 	for i := 0; i < indexShards; i++ {
 		shards[i].idx = map[model.LabelName]map[model.LabelValue][]model.Fingerprint{}
 	}
-	return &invertedIndex{
+	return &InvertedIndex{
 		shards: shards,
 	}
 }
 
-func (ii *invertedIndex) add(metric labelPairs, fp model.Fingerprint) {
-	i := &ii.shards[hashFP(fp)%indexShards]
-	i.mtx.Lock()
-	defer i.mtx.Unlock()
+// Add a fingerprint under the specified labels.
+func (ii *InvertedIndex) Add(labels []client.LabelPair, fp model.Fingerprint) {
+	shard := &ii.shards[hashFP(fp)%indexShards]
+	shard.add(labels, fp)
+}
+
+// Lookup all fingerprints for the provided matchers.
+func (ii *InvertedIndex) Lookup(matchers []*labels.Matcher) []model.Fingerprint {
+	if len(matchers) == 0 {
+		return nil
+	}
+
+	result := []model.Fingerprint{}
+	for i := range ii.shards {
+		fps := ii.shards[i].lookup(matchers)
+		result = append(result, fps...)
+	}
+
+	sort.Sort(fingerprints(result))
+	return result
+}
+
+// Delete a fingerprint with the given label pairs.
+func (ii *InvertedIndex) Delete(labels []client.LabelPair, fp model.Fingerprint) {
+	shard := &ii.shards[hashFP(fp)%indexShards]
+	shard.delete(labels, fp)
+}
+
+// LabelNames returns all label names.
+func (ii *InvertedIndex) LabelNames() model.LabelNames {
+	results := make([]model.LabelNames, 0, indexShards)
+
+	for i := range ii.shards {
+		shardResult := ii.shards[i].labelNames()
+		results = append(results, shardResult)
+	}
+
+	return mergeLabelNameLists(results)
+}
+
+// LabelValues returns the values for the given label.
+func (ii *InvertedIndex) LabelValues(name model.LabelName) model.LabelValues {
+	results := make([]model.LabelValues, 0, indexShards)
+
+	for i := range ii.shards {
+		shardResult := ii.shards[i].labelValues(name)
+		results = append(results, shardResult)
+	}
+
+	return mergeLabelValueLists(results)
+}
+
+const cacheLineSize = 64
```

Comment:
I think this deserves a comment noting this is the prevalent value for Intel and AMD CPUs as-at 2018.
(since I had to look it up)

Reply:
Done!